### ISSUE TITLE
Docker container should reread internal_users.yml

### ISSUE THREAD

**Issue by untoldone** [opendistro-for-elasticsearch/security#25]
Currently `opendistro_security/securityconfig/internal_users.yml` is only read for password hashes on the initial load.

All the documentation to date relies docker containers getting destroyed in-between edits to this file. This is not an expected usage pattern in any commonly used open source project configuration.

The expectation would be that editing this file takes effect after a service restart (or live). At a minimum, I'd expect the documentation around this to be much clearer that the container actually needs to be destroyed for this to take.

Also useful would be a method (or documentation if it already exists) on how to change the admin password after a cluster has been brought up for docker without destroying or stopping the container and related volumes as the docs I've seen seem for outside of docker seem to use tools not present in the docker container.

I would imagine changing a password without downtime would be a common maintenance task for production systems.

**Comment by shivdoshi** [comment]
You can reload the configuration stored in your `internal_users.yml` file and other config files by using the `securityadmin` tool. This ships by-default with the plugin and is located in the tools directory in the plugin. You can find more information and related commands in our documentation here - https://opendistro.github.io/for-elasticsearch-docs/docs/security/security-admin/

The `-rl` flag in following example reload the configuration files:
>./tools/securityadmin.sh  -icl -key "/usr/share/elasticsearch/config/kirk-key.pem" -cert "/usr/share/elasticsearch/config/kirk.pem" -cacert "/usr/share/elasticsearch/config/root-ca.pem" -nhnv -rl



>

**Comment by untoldone** [comment]
@shivdoshi Personally don't think this is sufficiently usable and still feel if nothing else its a real usability bug. To be totally honest, given this and several other odd / unexpected configuration issues I've decided we can't use opendistro just as feedback.

Also maybe I missed it, but I couldn't find these tools in the docker container.

**Comment by shivdoshi** [comment]
@untoldone If you want to update the `admin` password from Kibana/REST API, you can do that by removing the `readonly` flag for `admin` user from `internal_user.yml` file. We preferred the `securityadmin` approach so that changing admin password can only be done if the user has SSH access to the cluster. You can find more discussion on this here #10 .

The `securityadmin.sh` tool should be located under tools folder in the security plugin directory:`{$ES_HOME}/plugins/opendistro_security/tools/securityadmin.sh`. Please let me know if that is not the case for your docker image.

**Comment by untoldone** [comment]
Ah -- I must have miss-written my find, thanks for sharing the location. Not super important, but it was not properly chmod-ed to allow execution.

Again, I do still think this bug is worth keeping as I do think this is a current usability issue for someone trying to run this as a container in production -- its not super easy to set it up to launch safely in a production environment without the default password.

It is more typical to allow setting of passwords in environment variables such as host Postgres does it (https://hub.docker.com/_/postgres) by POSTGRES_PASSWORD or allow the loading of a password from a mounted secret file such as MySQL (https://hub.docker.com/_/mysql) with MYSQL_ROOT_PASSWORD_FILE.

**Comment by alolita** [comment]
@untoldone Good point - making it easier to launch safely in a prod environment is important. Categorizing this issue as a bug.

**Comment by mporracindie** [comment]
The problem with @shivdoshi solutions is that everytime you run securityadmin.sh you lose all your roles, mapping and users configurations.
There should be a easier way to change this passwords.

**Comment by sujithvm** [comment]
@untoldone There is an account API to change password for non-reserved users.  https://opendistro.github.io/for-elasticsearch-docs/docs/security-access-control/api/#change-password . For reserved users, you need to update password via super-admin

**Comment by untoldone** [comment]
@sujithvm I no longer recall exactly the nature of this issue. I also don't know what improvements you've made to the container since I last looked.

This said, I believe the issue was the default behavior didn't match other common pattern's used in docker such as allowing the default password to be set via an environment variable. I'm guessing what you're describing is a way to change the password after the docker container was created? If this is the case, that would increase the complexity of administering the service if run within docker as most docker orchestration platforms don't give you tools to execute commands within the context of the container while it is running -- everything must be done by files, and environment variables ahead of launching the container.

Again, I'm not entirely sure if this was exactly what was going on as its been a while since I tried using this container.
